.github/workflows/ci.yml

```diff
@@ -40,11 +40,83 @@ jobs:
       - name: Build project
         run: cargo build --workspace
 
+      - name: Restore cargo config
+        run: |
+          if [ -f .cargo/config.toml.bak ]; then
+            mv .cargo/config.toml.bak .cargo/config.toml
+          fi
+
+  test:
+    runs-on: macos-latest
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Cache dependencies
+        uses: actions/cache@v4
+        with:
+          path: |
+            ~/.cargo/registry
+            ~/.cargo/git
+            target
+          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-cargo-
+
+      - name: Install Rust nightly toolchain
+        uses: dtolnay/rust-toolchain@nightly
+        with:
+          components: rust-src
+
+      - name: Temporarily remove cargo config for test
+        run: |
+          if [ -f .cargo/config.toml ]; then
+            mv .cargo/config.toml .cargo/config.toml.bak
+          fi
+
       - name: Run tests
         run: cargo test --workspace
 
       - name: Restore cargo config
         run: |
           if [ -f .cargo/config.toml.bak ]; then
             mv .cargo/config.toml.bak .cargo/config.toml
-          fi
\ No newline at end of file
+          fi
+
+  cargo-pio-build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Cache Docker layers
+        uses: actions/cache@v4
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-docker-${{ hashFiles('Dockerfile') }}
+          restore-keys: |
+            ${{ runner.os }}-docker-
+
+      - name: Cache build artifacts
+        uses: actions/cache@v4
+        with:
+          path: |
+            target
+            .pio
+            .embuild
+          key: ${{ runner.os }}-cargo-pio-${{ hashFiles('**/Cargo.lock', 
+            'platformio.ini') }}
+          restore-keys: |
+            ${{ runner.os }}-cargo-pio-
+
+      - name: Build Docker image
+        run: |
+          docker build -t u701 .
+
+      - name: Run cargo pio build in Docker container
+        run: |
+          docker run --rm -v $(pwd):/app -w /app u701 bash -c "
+            wget -O /tmp/espup https://github.com/esp-rs/espup/releases/download/v0.14.0/espup-x86_64-unknown-linux-gnu && \
+            chmod +x /tmp/espup && \
+            mv /tmp/espup /usr/local/bin/espup && \
+            espup install && \
+            . /root/.espup/export-esp.sh && \
+            cargo pio build"
```

Comment:
The Docker container should already be able to run pio build.
Allow for `docker run --rm -v $(pwd):/app -w /app u701 cargo pio build`